diff --git a/compiler/GHC/Unit/Finder.hs b/compiler/GHC/Unit/Finder.hs
index cb91559cfe..13c477b6da 100644
--- a/compiler/GHC/Unit/Finder.hs
+++ b/compiler/GHC/Unit/Finder.hs
@@ -42,6 +42,11 @@
 import GHC.Unit.Env
 import GHC.Unit.Types
 import GHC.Unit.Module
+import GHC.Unit.Module.Graph
+       (
+           HomeModuleNameProvidersMap,
+           mgHomeModuleNameProvidersMap
+       )
 import GHC.Unit.Home
 import GHC.Unit.Home.Graph (UnitEnvGraph)
 import qualified GHC.Unit.Home.Graph as HUG
@@ -69,8 +74,13 @@
 import GHC.Driver.Config.Finder
 import qualified Data.Set as Set
 import Data.List.NonEmpty ( NonEmpty (..) )
+import Data.Set (Set)
+import qualified Data.Set as Set (empty, intersection, difference, null, toList)
+import qualified System.Directory as SD
 import qualified System.OsPath as OsPath
 import qualified Data.List.NonEmpty as NE
+import GHC.Types.Unique.Map
+import GHC.Types.Unique.Set
 
 type FileExt = OsString -- Filename extension
 type BaseName = OsPath  -- Basename of file
@@ -151,12 +161,13 @@
 
 findImportedModule :: HscEnv -> ModuleName -> PkgQual -> IO FindResult
 findImportedModule hsc_env mod pkg_qual =
-  let fc        = hsc_FC hsc_env
-      mhome_unit = hsc_home_unit_maybe hsc_env
-      dflags    = hsc_dflags hsc_env
-      fopts     = initFinderOpts dflags
+  let fc           = hsc_FC hsc_env
+      mb_home_unit = hsc_home_unit_maybe hsc_env
+      dflags       = hsc_dflags hsc_env
+      fopts        = initFinderOpts dflags
   in do
-    findImportedModuleNoHsc fc fopts (hsc_unit_env hsc_env) mhome_unit mod pkg_qual
+    let home_module_name_providers_map = mgHomeModuleNameProvidersMap (hsc_mod_graph hsc_env)
+    findImportedModuleNoHsc fc fopts (hsc_unit_env hsc_env) home_module_name_providers_map mb_home_unit mod pkg_qual
 
 findImportedModuleWithIsBoot :: HscEnv -> ModuleName -> IsBootInterface -> PkgQual -> IO FindResult
 findImportedModuleWithIsBoot hsc_env mod is_boot pkg_qual = do
@@ -169,55 +180,118 @@
   :: FinderCache
   -> FinderOpts
   -> UnitEnv
+  -> HomeModuleNameProvidersMap
   -> Maybe HomeUnit
   -> ModuleName
   -> PkgQual
   -> IO FindResult
-findImportedModuleNoHsc fc fopts ue mhome_unit mod_name mb_pkg =
+findImportedModuleNoHsc fc fopts ue home_module_name_providers_map mb_home_unit mod_name mb_pkg =
   case mb_pkg of
     NoPkgQual  -> unqual_import
-    ThisPkg uid | (homeUnitId <$> mhome_unit) == Just uid -> home_import
+    ThisPkg uid | (homeUnitId <$> mb_home_unit) == Just uid -> home_import
                 | Just os <- lookup uid other_fopts -> home_pkg_import (uid, os)
-                | otherwise -> pprPanic "findImportModule" (ppr mod_name $$ ppr mb_pkg $$ ppr (homeUnitId <$> mhome_unit) $$ ppr uid $$ ppr (map fst all_opts))
+                | otherwise -> pprPanic "findImportModule" (ppr mod_name $$ ppr mb_pkg $$ ppr (homeUnitId <$> mb_home_unit) $$ ppr uid $$ ppr (map fst all_opts))
     OtherPkg _ -> pkg_import
   where
-    all_opts = case mhome_unit of
-                Nothing -> other_fopts
-                Just home_unit -> (homeUnitId home_unit, fopts) : other_fopts
 
+    mb_home_unit_id :: Maybe UnitId
+    mb_home_unit_id = homeUnitId <$> mb_home_unit
 
-    home_import = case mhome_unit of
-                   Just home_unit -> findHomeModule fc fopts home_unit mod_name
-                   Nothing -> pure $ NoPackage (panic "findImportedModule: no home-unit")
+    all_opts :: [(UnitId, FinderOpts)]
+    all_opts = case mb_home_unit_id of
+        Nothing           -> other_fopts
+        Just home_unit_id -> (home_unit_id, fopts) : other_fopts
 
+    home_import :: IO FindResult
+    home_import = case mb_home_unit of
+        Just home_unit -> findHomeModule fc fopts home_unit mod_name
+        Nothing        -> pure $
+                          NoPackage (panic "findImportedModule: no home-unit")
 
+    {-
+        If the module is reexported, then look for it as if it was from the
+        perspective of that package which reexports it.
+    -}
+    home_pkg_import :: (UnitId, FinderOpts) -> IO FindResult
     home_pkg_import (uid, opts)
-      -- If the module is reexported, then look for it as if it was from the perspective
-      -- of that package which reexports it.
-      | Just real_mod_name <- mod_name `M.lookup` finder_reexportedModules opts =
-        findImportedModuleNoHsc fc opts ue (Just $ DefiniteHomeUnit uid Nothing) real_mod_name NoPkgQual
-      | mod_name `Set.member` finder_hiddenModules opts =
-        return (mkHomeHidden uid)
-      | otherwise =
-        findHomePackageModule fc opts uid mod_name
-
-    -- Do not be smart and change this to `foldr orIfNotFound home_import hs` as
-    -- that is not the same!! home_import is first because we need to look within ourselves
-    -- first before looking at the packages in order.
-    any_home_import = foldr1 orIfNotFound (home_import:| map home_pkg_import other_fopts)
-
-    pkg_import    = findExposedPackageModule fc fopts units  mod_name mb_pkg
-
-    unqual_import = any_home_import
-                    `orIfNotFound`
-                    findExposedPackageModule fc fopts units mod_name NoPkgQual
-
-    units     = case mhome_unit of
-                  Nothing -> ue_homeUnitState ue
-                  Just home_unit -> HUG.homeUnitEnv_units $ ue_findHomeUnitEnv (homeUnitId home_unit) ue
-    hpt_deps :: [UnitId]
-    hpt_deps  = homeUnitDepends units
-    other_fopts  = map (\uid -> (uid, initFinderOpts (homeUnitEnv_dflags (ue_findHomeUnitEnv uid ue)))) hpt_deps
+        | Just real_mod_name
+              <- flip M.lookup (finder_reexportedModules opts) mod_name
+            = findImportedModuleNoHsc fc opts ue home_module_name_providers_map
+                  (Just $ DefiniteHomeUnit uid Nothing)
+                  real_mod_name
+                  NoPkgQual
+        | Set.member mod_name (finder_hiddenModules opts)
+            = return (mkHomeHidden uid)
+        | otherwise
+            = findHomePackageModule fc opts uid mod_name
+
+    any_home_import :: IO FindResult
+    any_home_import = foldr1 orIfNotFound $
+                      home_import :| map home_pkg_import other_fopts
+    {-
+        Do not try to be smart and change this to `foldr orIfNotFound
+        home_import (map home_pkg_import other_fopts)`, as that would not be the
+        same. `home_import` is first because we need to first look within the
+        current unit before looking at the other units in order.
+    -}
+
+    pkg_import :: IO FindResult
+    pkg_import = findExposedPackageModule fc fopts unit_state mod_name mb_pkg
+
+    unqual_import :: IO FindResult
+    unqual_import
+        = any_home_import
+          `orIfNotFound`
+          findExposedPackageModule fc fopts unit_state mod_name NoPkgQual
+
+    unit_state :: UnitState
+    unit_state = case mb_home_unit_id of
+        Nothing           -> ue_homeUnitState ue
+        Just home_unit_id -> HUG.homeUnitEnv_units $
+                             ue_findHomeUnitEnv home_unit_id ue
+
+    home_unit_deps :: Set UnitId
+    home_unit_deps = Set.fromList (homeUnitDepends unit_state)
+
+    ranked_home_unit_deps :: [UnitId]
+    ranked_home_unit_deps = rankedHomeUnitDeps home_module_name_providers_map
+                                               mod_name
+                                               home_unit_deps
+
+    other_fopts :: [(UnitId, FinderOpts)]
+    other_fopts
+        = [
+              (uid, opts) |
+                  uid <- ranked_home_unit_deps,
+                  let opts = initFinderOpts $
+                             homeUnitEnv_dflags (ue_findHomeUnitEnv uid ue)
+          ]
+
+rankedHomeUnitDeps :: HomeModuleNameProvidersMap
+                   -> ModuleName
+                   -> Set UnitId
+                   -> [UnitId]
+rankedHomeUnitDeps _ _ home_unit_deps | Set.null home_unit_deps
+    = []
+rankedHomeUnitDeps home_module_name_providers_map mod_name home_unit_deps
+    = Set.toList cached_deps ++ Set.toList uncached_deps
+    where
+
+    cached_providers :: Set UnitId
+    cached_providers = lookupWithDefaultUniqMap home_module_name_providers_map
+                                                Set.empty
+                                                mod_name
+
+    cached_deps :: Set UnitId
+    cached_deps = Set.intersection home_unit_deps cached_providers
+
+    uncached_deps :: Set UnitId
+    uncached_deps = Set.difference home_unit_deps cached_providers
+{-
+    The special handling of the situation where the dependency set is empty does
+    not change the result, but it avoids triggering evaluation of the module
+    graph.
+-}
 
 -- | Locate a plugin module requested by the user, for a compiler
 -- plugin.  This consults the same set of exposed packages as
@@ -235,15 +309,15 @@
 findPluginModule hsc_env mod_name = do
   let fc = hsc_FC hsc_env
   let units = hsc_units hsc_env
-  let mhome_unit = hsc_home_unit_maybe hsc_env
-  findPluginModuleNoHsc fc (initFinderOpts (hsc_dflags hsc_env)) units mhome_unit mod_name
+  let mb_home_unit = hsc_home_unit_maybe hsc_env
+  findPluginModuleNoHsc fc (initFinderOpts (hsc_dflags hsc_env)) units mb_home_unit mod_name
 
 
 -- | A version of findExactModule which takes the exact parts of the HscEnv it needs
 -- directly.
 findExactModuleNoHsc :: FinderCache -> FinderOpts -> UnitEnvGraph FinderOpts -> UnitState -> Maybe HomeUnit -> InstalledModule -> IsBootInterface -> IO InstalledFindResult
-findExactModuleNoHsc fc fopts other_fopts unit_state mhome_unit mod is_boot = do
-  res <- case mhome_unit of
+findExactModuleNoHsc fc fopts other_fopts unit_state mb_home_unit mod is_boot = do
+  res <- case mb_home_unit of
     Just home_unit
      | isHomeInstalledModule home_unit mod
         -> findInstalledHomeModule fc fopts (homeUnitId home_unit) (moduleName mod)
diff --git a/compiler/GHC/Unit/Module/Graph.hs b/compiler/GHC/Unit/Module/Graph.hs
index e56ee33ca7..185164d802 100644
--- a/compiler/GHC/Unit/Module/Graph.hs
+++ b/compiler/GHC/Unit/Module/Graph.hs
@@ -69,6 +69,8 @@
    , mgLookupModule
    , mgLookupModuleName
    , mgHasHoles
+   , HomeModuleNameProvidersMap
+   , mgHomeModuleNameProvidersMap
    , showModMsg
 
     -- ** Reachability queries
@@ -158,10 +160,12 @@
 import GHC.Utils.Misc ( partitionWith )
 
 import System.FilePath
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Map (Map)
 import qualified Data.Map as Map
 import GHC.Types.Unique.DSet
-import qualified Data.Set as Set
-import Data.Set (Set)
+import GHC.Types.Unique.Map (UniqMap, emptyUniqMap, listToUniqMap_C)
 import GHC.Unit.Module
 import GHC.Unit.Module.ModNodeKey
 import GHC.Unit.Module.Stage
@@ -206,14 +210,32 @@
   -- Cached computation, whether any of the ModuleGraphNode are isHoleModule,
   -- This is only used for a hack in GHC.Iface.Load to do with backpack, please
   -- remove this at the earliest opportunity.
+  , mg_home_module_name_providers_map :: HomeModuleNameProvidersMap
+    -- ^ For each module name, which home units provide it.
   }
 
+type HomeModuleNameProvidersMap = UniqMap ModuleName (Set UnitId)
+
+mkHomeModuleNameProvidersMap :: [ModuleGraphNode] -> HomeModuleNameProvidersMap
+mkHomeModuleNameProvidersMap nodes
+    = listToUniqMap_C Set.union $
+      [
+          (moduleName, Set.singleton unitID) |
+              ModuleNode _ moduleNodeInfo <- nodes,
+              let moduleName = moduleNodeInfoModuleName moduleNodeInfo,
+              let unitID     = moduleNodeInfoUnitId moduleNodeInfo
+      ]
+
+mgHomeModuleNameProvidersMap :: ModuleGraph -> HomeModuleNameProvidersMap
+mgHomeModuleNameProvidersMap = mg_home_module_name_providers_map
+
 -- | Why do we ever need to construct empty graphs? Is it because of one shot mode?
 emptyMG :: ModuleGraph
 emptyMG = ModuleGraph [] (graphReachability emptyGraph, const Nothing)
                          (graphReachability emptyGraph, const Nothing)
                          (graphReachability emptyGraph, const Nothing)
                          False
+                         emptyUniqMap
 
 -- | Construct a module graph. This function should be the only entry point for
 -- building a 'ModuleGraph', since it is supposed to be built once and never modified.
@@ -312,7 +334,7 @@
   where
     duplicate_errs = rights (Map.elems node_types)
 
-    node_types :: Map.Map NodeKey (Either ModuleNodeType ModuleGraphInvariantError)
+    node_types :: Map NodeKey (Either ModuleNodeType ModuleGraphInvariantError)
     node_types = Map.fromListWithKey go [ (mkNodeKey n, Left (moduleNodeType n)) | n <- mg_mss ]
       where
         -- Multiple nodes with the same key are not allowed.
@@ -323,7 +345,7 @@
 
 -- | Check that all dependencies in the graph are present in the node_types map.
 -- This is a helper function used by checkModuleGraph.
-checkAllDependenciesInGraph :: Map.Map NodeKey (Either ModuleNodeType ModuleGraphInvariantError)
+checkAllDependenciesInGraph :: Map NodeKey (Either ModuleNodeType ModuleGraphInvariantError)
                             -> ModuleGraphNode
                             -> Maybe ModuleGraphInvariantError
 checkAllDependenciesInGraph node_types node =
@@ -338,7 +360,7 @@
 -- | Check if for the fixed module node invariant:
 --
 --   Fixed nodes can only depend on other fixed nodes.
-checkFixedModuleInvariant :: Map.Map NodeKey (Either ModuleNodeType ModuleGraphInvariantError)
+checkFixedModuleInvariant :: Map NodeKey (Either ModuleNodeType ModuleGraphInvariantError)
                 -> ModuleGraphNode
                 -> Maybe ModuleGraphInvariantError
 checkFixedModuleInvariant node_types node = case node of
@@ -488,13 +510,17 @@
 -- To preserve invariants, 'f' can't change the isBoot status.
 mapMG :: (ModSummary -> ModSummary) -> ModuleGraph -> ModuleGraph
 mapMG f mg@ModuleGraph{..} = mg
-  { mg_mss = flip fmap mg_mss $ \case
-      InstantiationNode uid iuid -> InstantiationNode uid iuid
-      LinkNode uid nks -> LinkNode uid nks
-      ModuleNode deps (ModuleNodeFixed key loc)  -> ModuleNode deps (ModuleNodeFixed key loc)
-      ModuleNode deps (ModuleNodeCompile ms) -> ModuleNode deps (ModuleNodeCompile (f ms))
-      UnitNode deps uid -> UnitNode deps uid
+  { mg_mss = new_mss
+  , mg_home_module_name_providers_map = mkHomeModuleNameProvidersMap new_mss
   }
+  where
+    new_mss =
+      flip fmap mg_mss $ \case
+        InstantiationNode uid iuid -> InstantiationNode uid iuid
+        LinkNode uid nks -> LinkNode uid nks
+        ModuleNode deps (ModuleNodeFixed key loc)  -> ModuleNode deps (ModuleNodeFixed key loc)
+        ModuleNode deps (ModuleNodeCompile ms) -> ModuleNode deps (ModuleNodeCompile (f ms))
+        UnitNode deps uid -> UnitNode deps uid
 
 -- | Map a function 'f' over all the 'ModSummaries', in 'IO'.
 -- To preserve invariants, 'f' can't change the isBoot status.
@@ -847,7 +873,7 @@
 -- described in the export list haddocks.
 --------------------------------------------------------------------------------
 
-newtype NodeMap a = NodeMap { unNodeMap :: Map.Map NodeKey a }
+newtype NodeMap a = NodeMap { unNodeMap :: Map NodeKey a }
   deriving (Functor, Traversable, Foldable)
 
 -- | Transitive dependencies, including SOURCE edges
@@ -923,7 +949,7 @@
     lookup_key :: ZeroScopeKey -> Maybe Int
     lookup_key = fmap zeroSummaryNodeKey . lookup_node
 
-    node_map :: Map.Map ZeroScopeKey ZeroSummaryNode
+    node_map :: Map ZeroScopeKey ZeroSummaryNode
     node_map =
       Map.fromList [ (s, node)
                    | node <- nodes
@@ -1022,7 +1048,7 @@
     lookup_key ::  (NodeKey, ModuleStage) -> Maybe Int
     lookup_key = fmap stageSummaryNodeKey . lookup_node
 
-    node_map :: Map.Map (NodeKey, ModuleStage) StageSummaryNode
+    node_map :: Map (NodeKey, ModuleStage) StageSummaryNode
     node_map =
       Map.fromList [ (s, node)
                    | node <- nodes
@@ -1040,10 +1066,13 @@
 extendMG :: ModuleGraph -> ModuleGraphNode -> ModuleGraph
 extendMG ModuleGraph{..} node =
   ModuleGraph
-    { mg_mss = node : mg_mss
-    , mg_graph =  mkTransDeps (node : mg_mss)
-    , mg_loop_graph = mkTransLoopDeps (node : mg_mss)
-    , mg_zero_graph = mkTransZeroDeps (node : mg_mss)
+    { mg_mss = new_mss
+    , mg_graph =  mkTransDeps new_mss
+    , mg_loop_graph = mkTransLoopDeps new_mss
+    , mg_zero_graph = mkTransZeroDeps new_mss
     , mg_has_holes = mg_has_holes || maybe False isHsigFile (moduleNodeInfoHscSource =<< mgNodeIsModule node)
+    , mg_home_module_name_providers_map = mkHomeModuleNameProvidersMap new_mss
     }
+  where
+    new_mss = node : mg_mss
 
diff --git a/ghc/GHCi/UI.hs b/ghc/GHCi/UI.hs
index 0243b5767d..204a0e5169 100644
--- a/ghc/GHCi/UI.hs
+++ b/ghc/GHCi/UI.hs
@@ -478,9 +478,9 @@
 default_args :: [String]
 default_args = []
 
-interactiveUI :: GhciSettings -> [(FilePath, Maybe UnitId, Maybe Phase)] -> Maybe [String]
+interactiveUI :: GhciSettings -> DynFlags -> [(FilePath, Maybe UnitId, Maybe Phase)] -> Maybe [String]
               -> Ghc ()
-interactiveUI config srcs maybe_exprs = do
+interactiveUI config initDynFlags srcs maybe_exprs = do
    -- HACK! If we happen to get into an infinite loop (eg the user
    -- types 'let x=x in x' at the prompt), then the thread will block
    -- on a blackhole, and become unreachable during GC.  The GC will
@@ -496,7 +496,7 @@
     -- Initialise buffering for the *interpreted* I/O system
    (nobuffering, flush) <- runInternal initInterpBuffering
 
-   installInteractiveHomeUnits
+   installInteractiveHomeUnits initDynFlags
 
    -- Update the LogAction. Ensure we don't override the user's log action lest
    -- we break -ddump-json (#14078)
@@ -689,8 +689,8 @@
 -- Within GHCi, you can rely on this property.
 --
 -- For motivation and design, see Note [Multiple Home Units aware GHCi]
-installInteractiveHomeUnits :: GHC.GhcMonad m => m ()
-installInteractiveHomeUnits = do
+installInteractiveHomeUnits :: GHC.GhcMonad m => DynFlags -> m ()
+installInteractiveHomeUnits dflags = do
   logger <- getLogger
   hsc_env <- GHC.getSession
   -- The initial set of DynFlags used for interactive evaluation is the same
@@ -698,7 +698,7 @@
   -- * -XExtendedDefaultRules and
   -- * -XNoMonomorphismRestriction.
   -- See Note [Changing language extensions for interactive evaluation] #10857
-  dflags <- getDynFlags
+
   let
     dflags0' =
       (xopt_set_unlessExplSpec LangExt.ExtendedDefaultRules xopt_set) .
@@ -728,30 +728,31 @@
   dflagsPrompt <- GHC.normaliseInteractiveDynFlags logger $
     setHomeUnitId interactiveGhciUnitId $ dflags0
       { packageFlags =
-        [ sessionUnitExposedFlag ] ++
-        [ homeUnitPkgFlag uid
-        | homeUnitEnv <- Foldable.toList $ hsc_HUG hsc_env
-        , Just homeUnit <- [homeUnitEnv_home_unit homeUnitEnv]
-        , let uid = homeUnitId homeUnit
-        ] ++
-        (packageFlags dflags0)
+        concat
+          [ [ sessionUnitExposedFlag ]
+          , [ homeUnitPkgFlag uid
+            | homeUnitEnv <- Foldable.toList $ hsc_HUG hsc_env
+            , Just homeUnit <- [homeUnitEnv_home_unit homeUnitEnv]
+            , let uid = homeUnitId homeUnit
+            ]
+          , packageFlags dflags0
+          ]
       , importPaths = []
       }
 
   let
     -- Explicitly depends on all current home units.
-    -- Additionally, we remove all 'importPaths', to avoid accidentally adding
-    -- any 'Target's to this 'Unit' that are not ':load'ed.
     dflagsSession =
       setHomeUnitId interactiveSessionUnitId $ dflags
         { packageFlags =
-          [ homeUnitPkgFlag uid
-          | homeUnitEnv <- Foldable.toList $ hsc_HUG hsc_env
-          , Just homeUnit <- [homeUnitEnv_home_unit homeUnitEnv]
-          , let uid = homeUnitId homeUnit
-          ] ++
-          (packageFlags dflags)
-        , importPaths = []
+          concat
+            [ [ homeUnitPkgFlag uid
+              | homeUnitEnv <- Foldable.toList $ hsc_HUG hsc_env
+              , Just homeUnit <- [homeUnitEnv_home_unit homeUnitEnv]
+              , let uid = homeUnitId homeUnit
+              ]
+            , packageFlags dflags
+            ]
         }
 
   let
@@ -2282,9 +2283,8 @@
     checkTargetModule :: GhciMonad m => ModuleName -> m Bool
     checkTargetModule m = do
       hsc_env <- GHC.getSession
-      let home_unit = hsc_home_unit hsc_env
       result <- liftIO $
-        Finder.findImportedModule hsc_env m (ThisPkg (homeUnitId home_unit))
+        Finder.findImportedModule hsc_env m NoPkgQual
       case result of
         Found _ _ -> return True
         _ -> do reportError (GhciModuleError $ GhciModuleNotFound (moduleNameString m))
diff --git a/ghc/Main.hs b/ghc/Main.hs
index 259678550f..3558741350 100644
--- a/ghc/Main.hs
+++ b/ghc/Main.hs
@@ -294,6 +294,7 @@
   throwGhcException (CmdLineError "not built for interactive use")
 #else
 ghciUI units srcs maybe_expr = do
+  initFlags <- getDynFlags
   hs_srcs <- case NE.nonEmpty units of
     Just ne_units -> do
       initMulti ne_units (checkOptions DoMake)
@@ -304,7 +305,7 @@
           s <- initMake srcs
           dflags <- getDynFlags
           return $ map (uncurry (,Just $ homeUnitId_ dflags,)) s
-  interactiveUI defaultGhciSettings hs_srcs maybe_expr
+  interactiveUI defaultGhciSettings initFlags hs_srcs maybe_expr
 #endif
 
 -- ----------------------------------------------------------------------------
diff --git a/testsuite/tests/ghci/prog023/src/A.hs b/testsuite/tests/ghci/prog023/src/A.hs
new file mode 100644
index 0000000000..0cc60c8fb5
--- /dev/null
+++ b/testsuite/tests/ghci/prog023/src/A.hs
@@ -0,0 +1,4 @@
+module A where
+
+foo :: String
+foo = "Called foo"
diff --git a/testsuite/tests/ghci/prog024/src/A.hs b/testsuite/tests/ghci/prog024/src/A.hs
new file mode 100644
index 0000000000..0cc60c8fb5
--- /dev/null
+++ b/testsuite/tests/ghci/prog024/src/A.hs
@@ -0,0 +1,4 @@
+module A where
+
+foo :: String
+foo = "Called foo"
diff --git a/testsuite/tests/ghci/prog024/src/B.hs b/testsuite/tests/ghci/prog024/src/B.hs
new file mode 100644
index 0000000000..0fb55899fd
--- /dev/null
+++ b/testsuite/tests/ghci/prog024/src/B.hs
@@ -0,0 +1,5 @@
+module B where
+
+import A
+
+bar = "Called bar"
